pkg/object/mutation/types.go

```diff
@@ -0,0 +1,151 @@
+	// +optional
+	Token string `json:"token,omitempty"`
+}
+
+// ResourceReference is a reference to a KRM resource by name and kind.
+// One of APIVersion or Group is required.
+// Group is generally preferred, to avoid needing to update the version in lock
+// step with the referenced resource.
+// If neither is provided, the empty group is used.
+type ResourceReference struct {
+	// Kind is a string value representing the REST resource this object represents.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	Kind string `json:"kind"`
+
+	// APIVersion defines the versioned schema of this representation of an object.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
+	// +optional
+	APIVersion string `json:"apiVersion,omitempty"`
+
+	// Group is accepted as a version-less alternative to APIVersion
+	// More info: https://kubernetes.io/docs/reference/using-api/#api-groups
+	// +optional
+	Group string `json:"group,omitempty"`
+
+	// Name of the resource.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string `json:"name,omitempty"`
+
+	// Namespace is optional, defaults to the namespace of the target resource.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// NewResourceReference returns the object as a ResourceReference
+func NewResourceReference(obj *unstructured.Unstructured) ResourceReference {
+	return ResourceReference{
+		Name:       obj.GetName(),
+		Namespace:  obj.GetNamespace(),
+		Kind:       obj.GetKind(),
+		APIVersion: obj.GetAPIVersion(),
+	}
+}
+
+// GroupVersionKind satisfies the ObjectKind interface for all objects that
+// embed TypeMeta. Prefers Group over APIVersion.
+func (r ResourceReference) GroupVersionKind() schema.GroupVersionKind {
+	if r.Group != "" {
+		return schema.GroupVersionKind{Group: r.Group, Kind: r.Kind}
+	}
+	return schema.FromAPIVersionAndKind(r.APIVersion, r.Kind)
+}
+
+// ObjMetadata returns the name, namespace, group, and kind of the ResourceReference
+func (r ResourceReference) ObjMetadata() object.ObjMetadata {
+	return object.ObjMetadata{
+		Name:      r.Name,
+		Namespace: r.Namespace,
+		GroupKind: r.GroupVersionKind().GroupKind(),
+	}
+}
+
+// String returns the format GROUP[/VERSION][/namespaces/NAMESPACE]/KIND/NAME
+func (r ResourceReference) String() string {
+	group := r.Group
+	if group == "" {
+		group = r.APIVersion
+	}
+	if r.Namespace != "" {
+		return fmt.Sprintf("%s/namespaces/%s/%s/%s", group, r.Namespace, r.Kind, r.Name)
+	}
+	return fmt.Sprintf("%s/%s/%s", group, r.Kind, r.Name)
+}
+
+// Equal returns true if the ResourceReference sets are eqivelent, ignoring version.
+// Fulfills Equal interface from github.com/google/go-cmp
+func (r ResourceReference) Equal(b ResourceReference) bool {
```

Comment:
`Equal` will mostly work when ignoring version. I think this will probably work for what it is being used for. The only true way to compare resources is by comparing their API Server assigned `UID`. For example, the following corner case will incorrectly say the two resources are not the same when they are:

Resource 1: 
Group/Version: `apiextensions.k8s.io/v1beta1`
Kind: `Deployment`
Namespace: `foo`,
Name: `bar`

Resource 2 (same resource):
Group/Version: `apps/v1`
Kind: `Deployment`
Namespace: `foo`,
Name: `bar`

Most of these corner cases have been deprecated by version 1.16 and later.

Reply:
The only two uses of Equal right now are for detecting self-references and test comparisons. In these cases, we don't really care about the api version.

But you're right that group mismatch would break the self-reference detection if the group changed between versions. This is an interesting edge case. I suspect that it would cause the resource to be treated like two separate resources, with one depending on the other, and both being added to the dependency graph. The graph doesn't look like it checks for duplicates. I'm not even really sure how we would do client-side deduplication when the group differs like this.